### PR TITLE
Add action to deploy Parse Cloud Code

### PR DESCRIPTION
**Pull request by teddynewell**
Adds an action to deploy Parse cloud code with optional application, release notes, and path to /Parse directory.

This action can be used for projects that use [Parse](https://parse.com) for backend infrastructure.

### CODE REVIEW

File: lib/assets/custom_action_template.rb [hunk 1]
```diff
@@ -25,12 +25,14 @@ def self.run(params)
       #####################################################
 
       def self.description
+        # Give a short description (<= 80 characters) for this action
```
Comment (KrauseFx) [inline lib/assets/custom_action_template.rb]: :+1:

File: lib/fastlane/actions/parse_deploy.rb [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+        return exit_code
+      end
+
+      #####################################################
+      # @!group Documentation
+      #####################################################
+
+      def self.description
+        "Deploy Parse Cloud code"
```
Comment (KrauseFx) [inline lib/fastlane/actions/parse_deploy.rb]: This should make clear it's about server code, and has nothing to do with push notifications maybe?

File: lib/fastlane/actions/parse_deploy.rb [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+      PARSE_DEPLOY_EXIT_STATUS = :PARSE_DEPLOY_EXIT_STATUS
+    end
+
+    class ParseDeployAction < Action
+      def self.run(params)
+        parse_application = params[:application]
+        parse_directory = params[:parse_directory]
+        release_notes = params[:release_notes]
+
+        if !parse_application || parse_application.empty?
```
Comment (KrauseFx) [inline lib/fastlane/actions/parse_deploy.rb]: Replace with 

```ruby
parse_application.to_s.length == 0
```

File: lib/fastlane/actions/parse_deploy.rb [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+      end
+
+      def self.available_options
+        [
+          FastlaneCore::ConfigItem.new(key: :application,
+                                       env_name: "FL_PARSE_DEPLOY_APPLICATION", # The name of the environment variable
+                                       description: "Target Application for ParseDeployAction", # a short description of this parameter
+                                       is_string: false, # true: verifies the input is a string, false: every kind of value
+                                       optional: true,
+                                       default_value: ""), # the default value if the user didn't provide one
```
Comment (KrauseFx) [inline lib/fastlane/actions/parse_deploy.rb]: You can just get rid of this line and the default_value will be `nil`

Comment (KrauseFx) [inline lib/fastlane/actions/parse_deploy.rb]: Oh I see you actually want an empty string, never mind then

**Comment by KrauseFx** [comment]
This is great, thank you for your contribution - I added some comments before it's ready to be merged.

Out of curiosity, how do you use it in your project? You deploy your iOS app and you back-end at the same time using `fastlane`?

**Comment by KrauseFx** [comment]
Please also update `Actions.md` with a short description on how to use this integration :+1:

File: lib/fastlane/actions/parse_deploy.rb (hunk 3 above)
Comment (teddynewell) [inline lib/fastlane/actions/parse_deploy.rb]: Replaced in 0faab33

File: lib/fastlane/actions/parse_deploy.rb (hunk 2 above)
Comment (teddynewell) [inline lib/fastlane/actions/parse_deploy.rb]: Clarified description and details in 484efc7

**Comment by teddynewell** [comment]
Updated `Actions.md` in aafd394

**Comment by teddynewell** [comment]
We just started using `fastlane` to manage our synchronized Development -> Staging -> Production deployment.  Before pushing nightly and TestFlight builds, we make sure any back-end changes are deployed to the appropriate environment.  It's working well for our bootstrap project.

Very happy with the tool suite so far – thanks for making it :smile:

File: docs/Actions.md [hunk 5]
```diff
@@ -774,6 +774,25 @@ appetize(
...
 ```
 
+### [Parse Cloud Code Deploy](https://www.parse.com/docs/cloudcode/guide#command-line-deploying)
+This will deploy server-side code to Parse Cloud from your local Parse installation.
+
+```ruby
+parse_deploy
+```
+
+You can pass all options listed in `parse_deploy --help` in `fastlane`:
```
Comment (KrauseFx) [inline docs/Actions.md]: Please remove this part, as it's described on the top of the document :+1:

File: lib/fastlane/actions/parse_deploy.rb [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        else
+          deploy_info = "#{parse_application}"
+        end
+        Helper.log.info "Deploying Parse cloud files to #{deploy_info} ⛅️".green
+
+        parse_path = File.expand_path(parse_directory)
+        exit_code = 1
+        if File.exist?(parse_path)
+          Dir.chdir(parse_path) do
+            command = "Parse deploy"
```
Comment (KrauseFx) [inline lib/fastlane/actions/parse_deploy.rb]: Shouldn't `Parse` be lower case?

Comment (KrauseFx) [inline lib/fastlane/actions/parse_deploy.rb]: What if this tool is not installed? Maybe we should add something like
```ruby
raise "Please install `parse` using `...`" unless `parse --help`.include?("something here")
```

File: lib/fastlane/actions/parse_deploy.rb [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+            if parse_application && !parse_application.empty?
+              command << " #{parse_application}"
+            end
+            if release_notes && !release_notes.empty?
+              command << " --description=\"#{release_notes}\""
+            end
+            Helper.log.info "Deploy command: #{command}".blue
```
Comment (KrauseFx) [inline lib/fastlane/actions/parse_deploy.rb]: Please change to `.yellow`, which is the default color for commands

**Comment by KrauseFx** [comment]
That's so great, thank you so much for this amazing pull request. I added a few more comments before it's ready to be merged, sorry for being so picky :wink:

File: docs/Actions.md (hunk 5 above)
Comment (teddynewell) [inline docs/Actions.md]: Removed in e2732a1

File: lib/fastlane/actions/parse_deploy.rb (hunk 6 above)
Comment (teddynewell) [inline lib/fastlane/actions/parse_deploy.rb]: Good thought – added in cdad8c6

File: lib/fastlane/actions/parse_deploy.rb (hunk 7 above)
Comment (teddynewell) [inline lib/fastlane/actions/parse_deploy.rb]: Changed in cdad8c6

**Comment by teddynewell** [comment]
Responded to your helpful comments – I'm not too familiar with Ruby, so I appreciate the suggestions.

**Comment by KrauseFx** [comment]
Ah, thanks @teddynewell. Somehow this ended up in a merge conflict. Sorry about the the trouble :wink:

**Comment by teddynewell** [comment]
Updated to fix the merge conflict :)

**Comment by KrauseFx** [comment]
@teddynewell thanks for updating the PR. One last thing before it's ready to be merged: could you please squash your commits to one? :+1:

**Comment by KrauseFx** [comment]
Hey @teddynewell, just wanted to check if you had time to squash the commits? Would love to have this included in fastlane :rocket:

**Comment by teddynewell** [comment]
Thanks for the followup – haven’t had time yet, as I spaced out on my clone and made changes on master.  I’ll figure it out and update the PR soon :)

**Comment by ohwutup** [comment]
hey @teddynewell! i am going to close this PR for now as it's been inactive for some time, but please feel free to open another PR when you get the time to do the final cleanup and squash!! thanks again!

**Comment by KrauseFx** [comment]
Thanks for updating this issue @ohwutup. Correct me if I'm wrong, but Parse will shutdown soon, so there is no real value in working on this, right?

**Comment by ohwutup** [comment]
oh, you're right. not sure why i didn't think of that, although, they did open source their platform right?

**Comment by teddynewell** [comment]
Thanks for closing this @ohwutup – I de-prioritized the last cleanup to the PR since the Parse shutdown until I can investigate if the open-source version has the same command line tool.  If it does, I'll create a new PR with the final squashing.

**Pull request closed without merging**